pkg/pcrbruteforcer/analyze_unexpected_log_entry.go

```diff
@@ -0,0 +1,211 @@
+		if chunk.ForceData != nil {
+			result = append(result, fmt.Sprintf(`{"ForceData": "0x%X"}`, chunk.ForceData))
+			continue
+		}
+		result = append(result, fmt.Sprintf(`{"Range": %s}`, chunk.Range))
+	}
+	return "{" + strings.Join(result, ", ") + "}"
+}
+
+func (e logEntryExplanation) String() string {
+	var details []string
+
+	switch {
+	case e.Measurement != nil:
+		details = append(details, fmt.Sprintf("reproduced the digest using measurement: %v", formatMeasurement(*e.Measurement)))
+	case e.EventDataParsed != nil:
+		if len(e.EventDataParsed.Ranges) > 0 {
+			details = append(details, fmt.Sprintf("mentioned byte ranges: %v", e.EventDataParsed.Ranges))
+		}
+		if len(e.EventDataParsed.FvGUIDs) > 0 {
+			details = append(details, fmt.Sprintf("mentioned UUIDs: %v", e.EventDataParsed.FvGUIDs))
+		}
+	}
+
+	if len(e.RelatedNodes) != 0 {
+		details = append(details, fmt.Sprintf("related UEFI nodes: %s", e.RelatedNodes))
+	}
+
+	if len(details) == 0 {
+		return "<unable to get any info>"
+	}
+
+	return strings.Join(details, "; ")
+}
+
+func (e logEntryExplanation) guessMeasurement(
+	expectedMeasurement *pcr.Measurement,
+	ev *tpmeventlog.Event,
+	image []byte,
+) *pcr.Measurement {
+	if e.EventDataParsed != nil {
+		m := e.guessMeasurementFromEventRanges(
+			expectedMeasurement,
+			ev,
+			image,
+		)
+		if m != nil {
+			return m
+		}
+	}
+
+	return e.guessMeasurementFromEventRawData(
+		expectedMeasurement,
+		ev,
+	)
+}
+
+func (e logEntryExplanation) guessMeasurementFromEventRawData(
+	expectedMeasurement *pcr.Measurement,
+	ev *tpmeventlog.Event,
+) *pcr.Measurement {
+	return tryMeasurement(expectedMeasurement, ev, nil, pcr.DataChunks{{ID: pcr.DataChunkIDUnknown, ForceData: ev.Data}})
+}
+
+func (e logEntryExplanation) guessMeasurementFromEventRanges(
+	expectedMeasurement *pcr.Measurement,
+	ev *tpmeventlog.Event,
+	image []byte,
+) *pcr.Measurement {
+	var chunks pcr.DataChunks
+	for _, r := range e.EventDataParsed.Ranges {
+		if isPhysAddr(r.Offset, uint64(len(image))) {
+			r.Offset -= PhysAddrBase - uint64(len(image))
+		}
+		chunkID := pcr.DataChunkIDUnknown
+		var forceData []byte
+		if expectedMeasurement != nil && len(expectedMeasurement.Data) >= len(chunks) {
+			chunkID = expectedMeasurement.Data[len(chunks)].ID
+			forceData = expectedMeasurement.Data[len(chunks)].ForceData
+		}
+		chunks = append(chunks, pcr.DataChunk{
+			ID:        chunkID,
+			Range:     r,
+			ForceData: forceData,
+		})
+	}
+	return tryMeasurement(expectedMeasurement, ev, image, chunks)
+}
+
+func tryMeasurement(
+	expectedMeasurement *pcr.Measurement,
+	ev *tpmeventlog.Event,
+	image []byte,
+	chunks pcr.DataChunks,
+) *pcr.Measurement {
+	measurementID := pcr.MeasurementIDUnknown
+	if expectedMeasurement != nil {
+		measurementID = expectedMeasurement.ID
+	}
+	measurement := &pcr.Measurement{
+		ID:   measurementID,
+		Data: chunks,
+	}
+	h, err := ev.Digest.HashAlgo.Hash()
+	if err != nil {
+		panic(err) // should never happen
```

Comment:
Can we avoid adding panics to the code? Even if they should never happen they turn up and kill the service.

Reply:
Considering it should never happen it seems unreasonable to make the signature of the function more complex and require to process errors (which will never happen) upper-side. But this is not a strong opinion, so feel free to insist ( then will fix :) )

Reply:
If it's never going to happen then can we not either ignore the error, or just return nil?

Given it looks like this could only fail if we fail to create a Hash object or calculate a digest (below) this shouldn't be affected by untrusted BIOS image input, and really shouldn't fail, so I'll not push hard for the removal of the panic.

Reply:
> If it's never going to happen then can we not either ignore the error, or just return nil?

Personally I hate silent errors even more than panics :)
This error should never happen by design. But since an error value is returned by this function and I have choice to:
* Process it everywhere expecting to never happen (and thus complicate code without reason).
* Ignore it.
* Panic.

My priorities are:
panic > process > ignore

If I'm wrong about "should never happen by design" then better to surface the error than silently ignore it. In future we should be able to add go-belt observability into the application and replace this `panic(err)` with harmless `errmon.ObserveError(err)`.